### PR TITLE
fix DSS performance

### PR DESCRIPTION
**Pull request by simonbyrne**
Fixes a key performance regression in ClimaCore 0.10. The compiler wasn't able to recurse through nested types for DSS.

- [x] Code follows the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) OR N/A.
- [x] Unit tests are included OR N/A.
- [x] Code is exercised in an integration test OR N/A.
- [x] Documentation has been added/updated OR N/A.

### CODE REVIEW

**Comment by charleskawczynski** [comment]
Looks like this did it, awesome!

Main:
```
BenchmarkTools.Trial: 14 samples with 1 evaluation.
--
  | Range (min … max):  349.499 ms … 379.547 ms  ┊ GC (min … max): 0.00% … 5.44%
  | Time  (median):     374.925 ms               ┊ GC (median):    5.36%
  | Time  (mean ± σ):   371.951 ms ±   9.121 ms  ┊ GC (mean ± σ):  4.26% ± 2.29%
  |  
  | █     ▃
  | ▇▁▁▁▁▁▁▁▁▇▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▇▁▁▇▁▁▁▇▁▁▁▁█▁▁▇▁▇█▁▇▇ ▁
  | 349 ms           Histogram: frequency by time          380 ms <
  |  
  | Memory estimate: 98.63 MiB, allocs estimate: 1974780.
```

This PR:
```
BenchmarkTools.Trial: 24 samples with 1 evaluation.
--
  | Range (min … max):  213.333 ms … 221.598 ms  ┊ GC (min … max): 0.00% … 0.00%
  | Time  (median):     214.456 ms               ┊ GC (median):    0.00%
  | Time  (mean ± σ):   214.824 ms ±   1.632 ms  ┊ GC (mean ± σ):  0.00% ± 0.00%
  |  
  | ▁ ▁ ▁▁█▁
  | ▆▁█▆█▆████▁▆▆▁▆▁▆▁▁▆▁▁▁▁▁▆▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▆ ▁
  | 213 ms           Histogram: frequency by time          222 ms <
  |  
  | Memory estimate: 1.50 MiB, allocs estimate: 14844.
```

**Comment by charleskawczynski** [comment]
We should probably create a new release after this since it does result in a decent speedup

**Review by charleskawczynski — APPROVED** [review]
🎉

File: src/RecursiveApply/RecursiveApply.jl [hunk 1]
```diff
@@ -41,8 +41,9 @@ Recursively apply `fn` to each element of `X`
 rmap(fn::F, X) where {F} = fn(X)
 rmap(fn::F, X, Y) where {F} = fn(X, Y)
 rmap(fn::F, X::Tuple) where {F} = tuplemap(x -> rmap(fn, x), X)
+rmap(fn, X::Tuple{}, Y::Tuple{}) = ()
 rmap(fn::F, X::Tuple, Y::Tuple) where {F} =
-    tuplemap((x, y) -> rmap(fn, x, y), X, Y)
+    (rmap(fn, first(X), first(Y)), rmap(fn, Base.tail(X), Base.tail(Y))...)
```
Comment (charleskawczynski) [inline src/RecursiveApply/RecursiveApply.jl:46]: This may help other operators / operations!

File: test/Operators/hybrid/dss_opt.jl [hunk 2]
```diff
@@ -0,0 +1,41 @@
+using Test
```
Comment (charleskawczynski) [inline test/Operators/hybrid/dss_opt.jl:1]: @simonbyrne, can we include this in `test/runtests.jl`?

Comment (simonbyrne) [reply test/Operators/hybrid/dss_opt.jl:1]: ah, i meant to

**Pull request merged**